### DISCUSSION TITLE
Not set cache when data is empty

### DISCUSSION THREAD

**Discussion by xstreamka** [yiisoft/yii2#19814] [Ideas]
Real life example:
There are API requests that sometimes fall, nothing comes in response.
```php
$result = Yii::$app->cache->getOrSet('key', function () {
	return ApiHelper::getData();
}, 60 * 60 * 12, $dependency);
```
And if we save empty data (ApiHelper::getData) in the cache, then we will receive emptiness until the cache expires.

I suggest adding an option to indicate that we don't want to store empty data in the cache.
Example in getOrSet() vendor/yiisoft/yii2/caching/Cache.php
I have added a new $saveEmpty argument which is enabled by default.

```php
public function getOrSet($key, $callable, $duration = null, $dependency = null, $saveEmpty = true) // added $saveEmpty
{
    if (($value = $this->get($key)) !== false) {
        return $value;
    }

    $value = call_user_func($callable, $this);

    // added $saveEmpty, data will not be saved if $value is empty.
    if (!$saveEmpty && empty($value)) {
        return $value;
    }

    if (!$this->set($key, $value, $duration, $dependency)) {
        Yii::warning('Failed to set cache value for key ' . json_encode($key), __METHOD__);
    }

    return $value;
}
```

In my project I made a temporary solution.
```php
namespace common\components;

class CacheComponent extends \yii\caching\FileCache
{
    /**
     * @param bool $deleteEmpty true if empty data should not be cached
     * @return false|mixed
     */
    public function getOrSet($key, $callable, $duration = null, $dependency = null, $deleteEmpty = false)
    {
        $value = parent::getOrSet($key, $callable, $duration, $dependency);

        if ($deleteEmpty && empty($value)) {
            parent::delete($key);
        }

        return $value;
    }
}
```

I will be glad to feedback

**Comment by samdark** [comment]
Hi. Thanks for idea. It is interesting but the fact is for some cases empty response is valid. For each API the definition of failure is different. In one case it is empty response, in another case it is HTTP status code.

Therefore, because of variety of cases, it is better to handle it specifically like you did.

**Reply by xstreamka** [reply]
I have nothing against it, I agree that an empty result is also a result.
The API example was probably not quite correct, it could be any request to the database or microservice.
The idea is that I know exactly what the data should be and should be cached only if it exists. Otherwise, there will be empty data for a long time.
For the right to select and need an additional optional argument.
Anyway, I just wanted to get the point across, thanks for the feedback!
